src/key_defs.h

```diff
@@ -25,6 +25,16 @@ typedef union Key_ {
   };
   uint16_t raw;
 
+  Key_() = default;
+
+  constexpr Key_(uint8_t __keyCode, uint8_t __flags)
+    : keyCode(__keyCode), flags(__flags) {
+  }
+
+  constexpr Key_(uint16_t __raw)
```

Comment:
Should this constructor be `explicit`?

Reply:
You seem to have good C++ knowledge. That's very nice.

On the one hand, I agree with your suggestion. The explicit keyword protects against undesired automatic type conversions. On the other hand, I suppose that if someone passes a single argument, she or he intents to pass raw data. By adding the explicit keyword, we would force users to cast literal constants to appropriate type, am I right? If so, we would cause extra work without much gain.

The only other use case I can imagine is that someone intents to pass only a keycode and wants the flags member to be initialized with a default value. That could be made possible by making the raw-value constructor explicit and adding a default value for the two-parameter constructor's flag parameter. That, however, would mean that the type of the first (only) argument would decide on the constructor that is going to be called. I don't feel too comfortable with such a solution. To me it seems quite error prone.

Reply:
> By adding the explicit keyword, we would force users to cast literal constants to appropriate type, am I right?

Well currently (with no constructors) there's no implicit conversion from an integer to a `Key` and it doesn't seem to cause any problems. So while you're right that you'd need to say `(Key)99` or `Key(99)` or `Key{99}` to explicitly convert an integer to a `Key`, that doesn't seem to be needed anywhere today.

By adding a non-`explicit` constructor you make it possible to do e.g.

````
void some_function(const Key&);
// ...
some_function(99);
````

If that's a desirable property, don't make the constructor explicit.

> That could be made possible by making the raw-value constructor explicit and adding a default value for the two-parameter constructor's flag parameter. That, however, would mean that the type of the first (only) argument would decide on the constructor that is going to be called.

Yes, and if you use `int` or some other type instead of exactly `uint16_t` or `uint8_t` it would be ambiguous and fail  to compile.

> I don't feel too comfortable with such a solution. To me it seems quite error prone.

I agree.

Reply:
If I got you right, you agree that if everybody expects the following
```cpp
Key(42) == Key(uint16_t(42)) == Key{ .raw = 42 } /* C++-20 ;-) */
```
we should be better off without an `explict` single argument constructor.

Reply:
The first two cases are explicit conversions anyway, so work the same whether the single argument constructor is `explicit` or not.

Let's forget I said anything and leave it non-explicit :-)

Reply:
> Let's forget I said anything and leave it non-explicit :-)

No, please, I find it important to care about such things. C++ features like the `explicit` keyword can help to prevent strange errors and, thus, can safe the community a lot of time and frustration.

> The first two cases are explicit conversions anyway

Just for my understanding: AFAIK, the literal `42` [is by default](http://www.cplusplus.com/doc/tutorial/constants/) an `int`  and does therefore require an implicit type conversion to become `uint16_t`. So `Key(42)` is not supposed to compile if
```cpp
explicit constexpr Key_(uint16_t __raw) ...
```
Am I missing something?

Reply:
> No, please, I find it important to care about such things.

OK sure, I just didn't want to be a pain if it wasn't welcome :-)

> does therefore require an implicit type conversion to become `uint16_t`

No, `explicit` doesn't do that.

tl;dr whether the `Key` constructor is `explicit` controls whether there are implicit conversions to `Key` but it can't stop implicit conversions from `int` to `uint16_t` (or any other conversions between the built-in integer types).

Longer version: 

You can never stop `int` implicitly converting to other integer types, that's an unavoidable feature of C and C++. Making the `Key` constructor `explicit` means that this doesn't work:

    uint16_t num = 1;
    Key k = num;  // error, requires implicit conversion from `uint16_t` to `Key`

Instead you must explicitly create a `Key` using one of these forms:

    Key k = Key(num);    // constructor notation
    Key k = (Key)num;    // cast notation
    Key k(num);      // direct-initialization
    Key k{num};      // direct-list-initialization i.e. using {} not ()

and you can't just use an integer where a `Key` is required:

    void some_function(const Key&);
    // ...
    uint16_t num = 42;
    some_function(num);   // error, requires implicit conversion from uint16_t to Key

But all these examples behave exactly the same if we use `int num` instead of `uint16_t num`, because an `int` can always implicitly convert to `uint16_t`.

So making the `Key` constructor `explicit` prevents integers implicitly converting to `Key` objects, but doesn't change _which_ integers are valid, i.e. `Key(42)` and `Key((uint16_t)42)` are identical. They both do direct-initialization of a `Key` from the value `(uint16_t)42` (the first one just implicitly converts `(int)42` to `(uint16_t)42`, and _then_ calls the `Key` constructor).

Reply:
> OK sure, I just didn't want to be a pain if it wasn't welcome :-)

Definitely not. Your above explanations and the rest of our discussion are __very__ helpful. Thanks.

Reply:
I was thinking about this one again. Now that I understand what's going on with the implicit type conversion allowed by the lack of the `explicit` specifier, I'm wondering: if it's desirable to allow those implicit conversions, is there a good reason not to allow implicit conversions the other way by defining the cast operator `operator uint16_t()` (or perhaps `operatory uint16_t&()`)?

If both implicit casts are allowed, that would remove the need for comparison operator overloads entirely, wouldn't it?

Reply:
No no no. If you allow implicit conversions in both directions you basically lose any type safety at all, and reasoning about the code can get much harder.

As I said originally, "Generally single-argument constructors should be `explicit` to avoid unwanted implicit conversions" i.e. implicit conversions are generally unwanted, and the source of bugs and confusion. Allowing them in one direction after carefully deciding it makes sense is OK, but that doesn't mean even more implicit conversions would be even better.

Reply:
I was just about to answer my own question. You'd still get some of the type safety: passing a parameter of type `Foo` to a function `f(Bar b)` still gives a compile error, even if both `Foo` and `Bar` allow implicit conversions to and from `int`, but a comparison operator will happily convert them both and allow the comparison.

We _would_ still get the ability to have a very small definition of two or more different types that are thin integer wrappers, and be able to define polymorphic functions like `f(Apple x)` and `f(Orange x)`, and assignments of one type to the other will still give a compilation error,  but those comparison operators…

Reply:
I'm still not convinced that it makes sense to allow implicit conversions at all, for the record.

Reply:
One of the main problems with implicit conversions to `uint16_t` would be that suddenly all kinds of un-type-safe expressions would compile:

````
Key n = Key_N;
Key lctrl = Key_LeftControl;
Key newfile = lctrl + n;  // use "Ctrl-N" to open a file
Key file1 = lctrl + 1;  // use "Ctrl-1" to switch to first file
````

If you're going to allow implicit conversion back to integers you might as well forget the `Key` abstraction completely and just use plain integers.

This topic is covered in detail in classic books like [C++ Coding Standards](http://www.gotw.ca/publications/c++cs.htm) and [More Effective C++](https://www.aristeia.com/books.html).

Some discussion in the C++ guidelines:
[Avoid conversion operators](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Ro-conversion)    
[By default, declare single-argument constructors explicit](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-explicit)

Reply:
That example also applies (and is more relevant to) implicit conversion _from_ integers _to_ `Key` objects, though (also, less relevant, `lctrl + n` is also not what you want). That's an argument for disallowing both implicit conversions.

Reply:
To make this more relevant, I'm thinking this may apply better to possible ways that key addressing might change. If we had two classes, each storing a single byte, and one of them represents the address of a keyswitch (`KeyAddr`), and the other represents the address of an LED (`LedAddr`), we might reasonably want a two functions called `setColor()`, one of which is used to set the LED color on a particular key, and one of which is meant to set the color of a particular LED. Consider that the indices for these two things are different, and some keyboards have LEDs that are not associated with particular keys.

Using the simplest approach – typedefs – doesn't work, because `KeyAddr` and `LedAddr` would actually be the same type.

Allowing implicit conversions in both directions would greatly simplify the code used to implement the two types, while making sure that the correct `setColor()` function gets called. It would allow programmers to make errors in comparison and arithmetic operations, however. That's a far cry from "lose any type safety at all", and it would mean that someone reading through the code wouldn't encounter pages of operator overloads.

I'm just not totally convinced that either option is strictly better than the other. I'm very skeptical of the idea that allowing the conversion in one direction is the best option, though. It's much more confusing if it appears to work in some situations, but not others than if it either always works or never works.

Reply:
It's not my code, and I don't really have a horse in this race. I'm happy to offer my opinionated advice, but I'm not going to fight over it :smile:

But I do think implicit conversions in both directions is a serous design flaw. There's a reason all the C++ experts who write books advise against implicit conversions, even though they appear tempting and superficially useful.